File: Leetcode/dfs_recursion_backtracking/508_MostFrequentSubtreeSum/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    Map<Integer, Integer> map = new HashMap<>();
    int maxCount = 0;
    public int[] findFrequentTreeSum(TreeNode root) {
        List<Integer> res = new ArrayList<>();
        dfs(root);
        for(int num: map.keySet()){
            if(map.get(num)==maxCount)
                res.add(num);
        }
        return res.stream().mapToInt(i->i).toArray();
    }
        private int dfs(TreeNode root){
            if(root==null) return 0;
            root.val += dfs(root.left)+dfs(root.right);
            int occurrence = map.getOrDefault(root.val, 0)+1;
            maxCount = Math.max(maxCount, occurrence);
            map.put(root.val, occurrence);
            return root.val;
        }
}

File: Leetcode/dfs_recursion_backtracking/508_MostFrequentSubtreeSum/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def findMode(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        count = collections.Counter()
        self.dfs(root, count)
        max_count = max(count.itervalues())
        return [k for k, v in count.iteritems() if v == max_count]

    def dfs(self, root, count):
        if not root: return
        self.dfs(root.left,count)
        count[root.val] += 1
        self.dfs(root.right, count)
